File: labs/typescript/example_cli_provider.ts
/**
 * CLI examples with provider support.
 *
 * Demonstrates the new CLI commands:
 *   porto config check [--provider ...]
 *   porto ident --country DE --weight 20 [--provider ...]
 *   porto restrict --country DE [--provider ...]
 *   porto calc --type letter_standard --country DE --weight 20 [--provider ...]
 */

import { execFileSync } from "node:child_process";

function runPorto(args: string[]): unknown {
  const cmd = "pnpm";
  const fullArgs = ["exec", "porto", ...args];
  console.log(`Running: ${cmd} ${fullArgs.join(" ")}`);
  const output = execFileSync(cmd, fullArgs, { encoding: "utf8" });
  return JSON.parse(output);
}

function main(): void {
  try {
    // 1. Config check (default provider)
    console.log("\n--- porto config check ---");
    const config = runPorto(["config", "check", "--json"]) as Record<string, unknown>;
    console.log(JSON.stringify(config, null, 2));

    // 2. Config check with explicit provider
    console.log("\n--- porto config check --provider swisspost ---");
    const configCh = runPorto(["config", "check", "--provider", "swisspost", "--json"]) as Record<
      string,
      unknown
    >;
    console.log(JSON.stringify(configCh, null, 2));

    // 3. Restrict (geo restrictions)
    console.log("\n--- porto restrict ---");
    const restrict = runPorto(["restrict", "--country", "DE", "--json"]) as Record<string, unknown>;
    console.log(JSON.stringify(restrict, null, 2));

    // 4. Restrict with explicit provider
    console.log("\n--- porto restrict --provider swisspost ---");
    const restrictCh = runPorto([
      "restrict",
      "--country",
      "CH",
      "--provider",
      "swisspost",
      "--json",
    ]) as Record<string, unknown>;
    console.log(JSON.stringify(restrictCh, null, 2));

    console.log("\n✅ All CLI commands OK");
  } catch (error) {
    console.error("CLI command failed");
    if (error instanceof Error) {
      console.error(error.message);
    } else {
      console.error(String(error));
    }
    process.exit(1);
  }
}

main();

File: labs/python/example_cli_provider.py
"""
CLI examples with provider support.

Demonstrates SDK CLI commands:
  porto config check [--provider ...]
  porto ident --country DE --weight 20 [--provider ...]
  porto restrict --country DE [--provider ...]
  porto calc --type letter_standard --country DE --weight 20 [--provider ...]
"""

from __future__ import annotations

import json
import subprocess
import sys
from typing import Any


def run_porto(args: list[str]) -> Any:
    full = ["porto", *args]
    print(f"Running: {' '.join(full)}")
    out = subprocess.check_output(full, text=True)
    return json.loads(out)


def main() -> int:
    try:
        print("\n--- porto config check ---")
        print(json.dumps(run_porto(["config", "check", "--json"]), indent=2))

        print("\n--- porto config check --provider swisspost ---")
        print(
            json.dumps(
                run_porto(["config", "check", "--provider", "swisspost", "--json"]),
                indent=2,
            )
        )

        print("\n--- porto restrict ---")
        print(json.dumps(run_porto(["restrict", "--country", "DE", "--json"]), indent=2))

        print("\n--- porto restrict --provider swisspost ---")
        print(
            json.dumps(
                run_porto(
                    [
                        "restrict",
                        "--country",
                        "CH",
                        "--provider",
                        "swisspost",
                        "--json",
                    ]
                ),
                indent=2,
            )
        )

        print("\n✅ All CLI commands OK")
        return 0
    except Exception as exc:  # pragma: no cover - lab script
        print("CLI command failed")
        print(str(exc))
        return 1


if __name__ == "__main__":
    sys.exit(main())
